### ISSUE TITLE
Improve API documentation

### ISSUE THREAD

**Issue by amichair** [chartist-js/chartist#761]
In a nutshell, the API documentation is incomplete, sometimes unclear and unorganized, not easily searchable as a reference, and generally frustrating. This is my first day looking at Chartist, so you can get a fresh perspective from a new user's point of view. See also issue #409, which was closed without actually fixing anything. I'll try to give some concrete examples of what is unclear and suggestions for improvement, but feel free to come up with better ideas and generally review the UX to make the project more accessible:
- The first thing one sees in the API page is the header "AUTO-SCALE-AXIS.JS" followed by the sub-header MODULE "CHARTIST.AUTOSCALEAXIS". There is no explanation of what a Chartist module is or how to use it, which is probably something you'd want to explain in an intro paragraph on the page. You can also add a short summary of the existing modules and what each does before starting with the first module on the list. Or at least group them in a way that conveys their meaning or category (e.g. axis stuff, chart types, utilities, etc.). Or order them non-arbitrarily.
- There's a heading labeled axis.js with nothing under it.
- In the module section itself it would be useful to have a CDN link (possibly the full html script tag) and source link so a user can quickly find it and get going using it or reviewing it (I'm assuming the source code is readable, well-written and documented with comments - I didn't check it out yet.)
- It is unclear what the difference is between the module headers (module file name?) and sub-headers (module name?) - there seems to be a one-to-one correspondence. Only the latter appears on the sidebar menu, so what's the larger filename header for?
- The module names (or file names? Or both?) are shown in uppercase. Is that really how they appear in the links and code? I'm guessing no. Anything that appears in the html/code should be presented in the correct case so it can be easily copied/pasted/searched.
- The first module says: "The following options are used by this axis in addition to the default axis options outlined in the axis configuration of the chart default settings." Why not link to wherever those default axis options are?
- "default options" and "default settings" seems to be used interchangeably (unless I haven't yet understood the difference). It's better to stick with consistent terminology, to make it more clear and more searchable.
- "the chart default settings" - searching the page for any part of that term (also using 'options' instead of 'settings') doesn't seem to come up with anything useful.
- The example snippet only shows a "var options =..." definition. It would be nice to see where and how to use this options object. Or should this be merged with the main options object? Or is this a sub-object of the main options object (under axisX, which is not mentioned here at all but seems to be related according to the text above)? Is the same variable name being reused for confusingly different purposes between the different examples? I'm still not sure, since I couldn't get the module to work yet, I think.
- Throughout the documentation (api, getting started, etc.) some code snippets are shown, and some hidden behind a 'show code' button. The inconsistency is confusing, and makes it more likely not to look at the collapsed ones which seem less important, even though they're not (e.g. default options, which double as the only place the available options are mentioned).
- Default options should not hide behind a 'show code' button. Although it's technically presented as code, this is actually the only place where the available options are specified (I'm assuming the defaults include a comprehensive list of possible options), and as such are a core part of the API specification. They're arguably the most important part of the API documentation, and aren't even searchable when collapsed.
- The 'show code' button should turn into 'hide code' when expanded.
- Some of the API sections say "Check the examples for a detailed list." or something equivalent. The detailed list of valid parameters is an integral part of the API and should be listed right then and there. Examples should show usage examples, not define the API contract.
- Some of the API sections have an 'Examples' sub-section. There's also a separate Examples page. Saying "Check the examples for a detailed list." is ambiguous, and again - there's no need to send users on a hunt (especially when search functionality doesn't work in collapsed snippets)... just link to wherever it is you want them to look.
- Examples within the API docs (as opposed to the examples page) should all have some title, comment, footnote or any other form of explanation of what the example is meant to do.

Well, that's just a start from a first-time user. I hope these points will be helpful in improving the API documentation so that newbies like myself can get up and running quickly rather than going around in circles trying to figure out where the info I need is hiding. I think a little bit of work here can go a long way towards making the first steps with Chartist less painful, an possibly even a pleasure (there certainly are libraries that achieve that).

Now I'll get back to trying to get the auto-scale-axis to work by googling for working examples elsewhere :-)

**Comment by amichair** [comment]
Two more:
- The module names in the sidebar menu don't link to the respective module sections (starting with a short description of the module) as they should.
- From what I currently understand, the axis modules need to be activated by setting the 'type' property under the 'axisX' option of the main options object to the constructor name such as 'Chartist.AutoScaleAxis'. I'm not sure yet this is correct, but if so it should be mentioned in each of the axis module sections where one would expect to see how the module is to be used. A working example should be added as well so it can be used with a simple copy+paste. Same goes for whatever other configuration options enable other types of modules (interpolation perhaps? just guessing).

In short, the documentation for every module, method, option etc. should include all information needed to make it work, not just bits and pieces scattered around everywhere. A working example for each would be extremely helpful.

**Comment by gionkunz** [comment]
Thanks for the deep summary! This will be valuable input for #565

**Comment by gionkunz** [comment]
I'm closing this issue, and we keep track with the reference available in #565
